lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -31,3 +33,45 @@ def get_index_prefix(config):
         return config.get("Indexing", "index_prefix")
     except (NoOptionError, NoSectionError):
         return ""
+
+
+def gen_month_range(prefix, index, start, end):
+    """
+    gen_month_range Construct a comma-separated list of index names
+    qualified by year and month suitable for use in the Elasticsearch
+    /_search query URI.
+
+    The month is incremented by 1 from "start" to "end"; for example,
+
+    gen_month_range('drb.', 'v4.run.', '2020-08', '2020-10') will result
+    in
+
+        'drb.v4.run.2020-08,drb.v4.run.2020-09,drb.v4.run.2020-10,'
+
+    This code is adapted from Javascript in the Pbench dashboard's
+    'moment_constants.js' getAllMonthsWithinRange function.
+
+    Args:
+        prefix ([string]): The Pbench server index prefix
+        index ([string]): The desired monthly index root
+        start ([datetime.datetime]): The start time
+        end ([datetime.datetime]): The end time
+
+    Returns:
+        [string]: A comma-separated list of month-qualified index names
+    """
+    monthResults = list()
+    queryString = ""
+    for m in rrule.rrule(rrule.MONTHLY, dtstart=start, until=end):
+        monthResults.append(m.strftime("%Y-%m"))
+
+    # TODO: hardcoding the index here is risky. We need a framework to
+    # help the web services understand index versions and template
+    # formats, probably by building a persistent database from the
+    # index template documents at startup. This is TBD.
+    for monthValue in monthResults:
+        if index == "v4.result-data.":
+            queryString += f"{prefix + index + monthValue}-*,"
+        else:
+            queryString += f"{prefix + index + monthValue},"
```

Comment:
Well, I _did_ write it, just not in this PR.

More importantly, the code comment is a completely different angle on this, in that hardcoding the index prefix requires additional maintenance over time when we change the version of indices. Theoretically, I would be doing that, for example, in this PR as I'm changing the Elasticsearch document templates. I'm "cheating" for convenience as we haven't released the previous changes I made for Elasticsearch V7, and another version update seems pointless.

Currently, the indexer reads the template documents from disk, and generates a map of index names (e.g., `result-data` or `run`) to version (e.g., in this example, `v4`). I'd like to move that map into persistent storage accessible throughout the server so that we could refer her to the current version of the `result-data` index document template.

And, yes, that mapping could also encode (and I believe does) whether the date-series of the index is monthly (`run`) or finer granularity (`result-data`) requiring the trailing `-*`, making this logic much more robust. Separating out the template is a very minor simplification that doesn't much interest me right now, in context; it does nothing to address the real concerns with this code. ;-)

Reply:
I'm not a fan of using four lines of code where two will do, especially if the two-line approach moves the code closer to making it easier or more straightforward to format the index name externally.